### PR TITLE
CON-1120: Edit Conclave Mail Doc

### PR DESCRIPTION
**Pull request by JohnJoser3**
I have made the changes to simplify and reorder the contents of the Conclave Mail doc. A couple of other docs which refer to the conclave mail doc were changed too. Please review.

### CODE REVIEW

File: docs/docs/mail.md [hunk 1]
```diff
@@ -1,345 +1,180 @@
...
+To avoid the host guessing information from message timing, you can send empty Mail items even when you have nothing 
+to say. For example, if an enclave is running an auction between users, and you wish to hide who won, the enclave 
+can send a "you won" or "you lost" message to every client.
+
+## Comparison to a classical public key infrastructure
+
+Conclave uses Conclave Mail to connect clients to enclaves because of the following reasons:
+
+1. Conclave Mail can work with a small [Trusted Computing Base (TCB)](https://en.wikipedia.org/wiki/Trusted_computing_base) 
+   when compared to HTTPS. A small TCB removes several common security problems at the origin.
```
Comment (ac101m) [inline docs/docs/mail.md:145]: Compared to HTTPS, Conclave mail has a relatively small [Trusted Computing Base (TCB)](...). The small TCB helps to harden enclaves against potential [zero-day](https://en.wikipedia.org/wiki/Zero-day_(computing)) exploits that may exist in dependency libraries.

File: docs/docs/mail.md [hunk 2]
```diff
@@ -1,345 +1,180 @@
...
+
+2. The certificate-based architecture of SSL/TLS doesn't gel with enclave-based computing. Conclave Mail uses the
```
Comment (ac101m) [inline docs/docs/mail.md:147]: "Doesn't gel" maybe "doesn't integrate cleanly with"

File: docs/docs/mail.md [hunk 3]
```diff
@@ -1,345 +1,180 @@
...
+   * Get other benefits like reusing MQ brokers, integrating with Corda flows, storing messages to databases, and 
+     supporting M-to-1 inbound message collection.
+
+
+3. Conclave Mail's approach to enclave restarts and enclave upgrades is more suited than a classical architecture 
+   that requires a database to ensure session persistence.
+
+4. The primary reason to use HTTPS/REST is the availability of tools and libraries. However, none of these 
+   tools or libraries support SGX remote attestation. You must modify or adjust these tools in complex ways,
+   invalidating most benefits.
```
Comment (ac101m) [inline docs/docs/mail.md:180]: The second sentence can be softened as follows:
"In order to make use of these preexisting tools for remote attestation, they must be used in unusual ways that they were not designed for, which introduces design complexities and reduces their usefulness."

**Review by ac101m — COMMENTED** [review]
Still not sure about the last section, @shamsasari do you think we still need this?

File: docs/docs/mail.md (hunk 3 above)
Comment (JohnJoser3) [reply docs/docs/mail.md]: Rewrote: "To use these preexisting tools for remote attestation, you must modify them in unusual ways that they were not designed for, which introduces design complexities and reduces their usefulness."

File: docs/docs/mail.md (hunk 2 above)
Comment (JohnJoser3) [reply docs/docs/mail.md]: Done.

File: docs/docs/mail.md (hunk 1 above)
Comment (JohnJoser3) [reply docs/docs/mail.md]: Done.

File: docs/docs/mail.md [hunk 4]
```diff
@@ -1,345 +1,180 @@
...
+   
+   If you use TLS, a client communicating with an enclave needs to extract a remote attestation from a 
+   pseudo-certificate. As Conclave Mail uses the Noise protocol, you can provide a byte array instead of a 
+   certificate as part of the handshake.
+
+   In Conclave, the remote attestation data is represented as an
+   [`EnclaveInstanceInfo`](api/-conclave%20-core/com.r3.conclave.common/-enclave-instance-info/index.html) object. The
+   Noise protocol and Conclave doesn't restrict how you get the remote attestation. For example, you can
+   send remote attestation to the client by returning it in another API, publishing it on a web server, and putting it 
+   into a distributed hash table, a network drive, or a message queue.
```
[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md]: Done.

File: docs/docs/mail.md [hunk 5]
```diff
@@ -1,345 +1,180 @@
...
+message headers. You can use message headers to implement non-data-processing tasks like usage tracking and 
+prioritization at the host's end. The Mail headers contain the following fields:
+
+1. The _topic_. You can use it to distinguish between different streams of Mail items from the same client. It's a 
+   string similar to an email subject. Topics are scoped per sender and are not global. Clients can send multiple 
+   streams of related Mail items using a different topic for each stream. To avoid replay attacks, you should never 
+   reuse a topic for an unrelated Mail item. Using a random UUID in a topic is good practice to prevent reuse.
+
+2. The _sequence number_. Every Mail item under a topic has a sequence number that starts from zero and increments 
+   by one. The enclave rejects messages if the sequence number is not in order. This ensures that the enclave
```
[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md]: Done.

File: docs/docs/mail.md [hunk 6]
```diff
@@ -1,345 +1,180 @@
...
-data that's unencrypted, but it's often useful for the client, host and enclave to collaborate in various ways related
-to storage and routing of data. Even when the host is untrusted it may still be useful for the client to send data
-that is readable by the host and enclave simultaneously, but which the host cannot tamper with. Inside the enclave
-you can be assured that the header fields contain the values set by the client, because they're checked before
+Conclave Mail's plain-text headers are authenticated and tamper-proof. Clients can use these headers to connect 
+messages to structure a conversation logically. Clients can also hold multiple conversations simultaneously using 
+message headers. You can use message headers to implement non-data-processing tasks like usage tracking and 
+prioritization at the host's end. The Mail headers contain the following fields:
+
+1. The _topic_. You can use it to distinguish between different streams of Mail items from the same client. It's a
```
[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md]: Rewrote: "This is a string you can use to distinguish between different streams of Mail items from the same client. It resembles an email's subject line."

File: docs/docs/mail.md [hunk 7]
```diff
@@ -1,345 +1,180 @@
...
 ### Authentication
 
-Mail allows a message to prove it came from the owner of a particular key, as well as being
-encrypted to a destination public key. Because of this it's easy to encrypt a reply to the sender,
-although how a public key is mapped to a physical computer is up to the app developer. This capability is useful
-for another reason: if an enclave has a notion of a user identifiable by public key, mail can be cryptographically
-authenticated as having come from that user. This avoids the need for complex user login processes: the login
-process can be handled outside of the enclave, without compromising security. The envelope is protected this way
-along with the encrypted body.
+Conclave Mail enables a message to prove that it came from the owner of a particular key. If a key can identify a
```
[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md]: Done

File: docs/docs/mail.md [hunk 8]
```diff
@@ -1,345 +1,180 @@
...
+   * Eliminate complexity from enclave code.
+   * Avoid hacks like pseudo-certificates.
+   * Enable enclaves to talk to each other even if they can't be loaded simultaneously.
+   * Move session management *and expiry* out of the enclave.
+   * Get other benefits like reusing MQ brokers, integrating with Corda flows, storing messages to databases, and 
+     supporting M-to-1 inbound message collection.
+
+
+3. Conclave Mail's approach to enclave restarts and enclave upgrades is more suited than a classical architecture 
+   that requires a database to ensure session persistence.
```
[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md:179]: Googled ```HTTPS needs "database" to ensure session persistence```.

I wonder if the existing content is referring to [Using a Database for Persistent Storage (JDBC Persistence)](https://docs.oracle.com/cd/E23943_01/web.1111/e13712/sessions.htm#WBAPP313).

[the remark this answers is not in the document]
Comment (ac101m) [reply docs/docs/mail.md:179]: Yeah, that does make sense to some extent if "classical architecture" is taken to mean the usual layout of a web program, web server + database etc. On reflection I think this goes back to when mail was used not just for messaging but for persistence. I think we can remove it.

[the remark this answers is not in the document]
Comment (JohnJoser3) [reply docs/docs/mail.md]: Removed.

**Review by ac101m — APPROVED** [review]
Looks good to me! Still not 100% sure about the last section, but it's better than it was and I don't feel we should get too hung up on it. LGTM

**Comment by ac101m** [comment]
> I just noticed that this has been raised against `release/1.3`! This needs to be raised against master! Please re-raise.

Never mind, disregard this comment!

File: docs/docs/mail.md [hunk 9]
```diff
@@ -1,345 +1,186 @@
...
+
+Conclave Mail provides various features that are useful when building secure applications.
 
 ### Encryption
 
-The encryption key used by the enclave is private to that enclave but stable across restarts and
-enclave upgrades. This means messages encrypted and delivered by older clients can still be decrypted. The format
-uses the respected [Noise protocol framework](https://noiseprotocol.org/) (with AES/GCM and SHA-256), which is
-the same cryptographic framework used by WhatsApp, the Linux kernel WireGuard protocol and I2P.
+The encryption key used by an enclave is private to that enclave. The encryption key is stable across system restarts
```
Comment (shamsasari) [inline docs/docs/mail.md:64]: The encrption key used for Mail by the enclave is no longer stable across system restarts. This was changed in 1.2, so this bit needs to change! Instead say that the enclave uses a random session key which changes each time the enclave restarts.

File: docs/docs/mail.md [hunk 10]
```diff
@@ -1,345 +1,186 @@
...
-so fake pseudo-certificates can end up harder to deploy and work with than a purely technical analysis would suggest.
-Noise is more modular - you can provide any arbitrary byte array as part of the handshake, which can thus 
-include a certificate (of any format) if you want that, but it isn't necessary.
-
-In Conclave the "certificate" is the remote attestation data represented as an [`EnclaveInstanceInfo`](api/-conclave%20-core/com.r3.conclave.common/-enclave-instance-info/index.html) object. 
-It will usually make sense to expose this to the client via some app-specific mechanism, for example, returning it
-in some other API, publishing it on a web server, a network drive, a message queue, even putting it into a 
-distributed hash table. Noise and by extension Conclave doesn't care how you get this data, only that you have it.
+The Conclave runtime verifies that the sequence numbers under a topic always increment by one before passing a Mail 
+item to the enclave. This feature ensures that the host can't reorder messages or drop Mail items.
```
Comment (shamsasari) [inline docs/docs/mail.md:127]: the host can't reorder or drop Mail messages.

File: docs/docs/mail.md [hunk 11]
```diff
@@ -1,345 +1,186 @@
...
 
-A mail is an authenticated byte array with an encrypted body and a cleartext envelope. The mail can be up to two
-gigabytes in size, however, it must at this time fit into memory when loaded all at once. In practice mails will
-usually be much smaller than that. 
+Conclave uses *Conclave Mail*, a communication mechanism that allows you to send encrypted messages to enclaves that 
+can be decrypted only by enclaves that you trust. Conclave Mail ensures secure communication between a client and an 
+enclave.
 
-## Mail features
+## How Conclave Mail works
```
Comment (shamsasari) [inline docs/docs/mail.md:7]: I'm not sure I agree with starting by describing how Mail works. There is a lot of technical detail here, which is not necessarily relevant for a Conclave user who wants to first understand what Mail is.

File: docs/docs/mail.md [hunk 12]
```diff
@@ -1,345 +1,186 @@
...
+
+To avoid the host guessing information from message timing, you can send empty Mail items even when you have nothing 
+to say. For example, if an enclave is running an auction between users, and you wish to hide who won, the enclave 
+can send a "you won" or "you lost" message to every client.
+
+## Comparison to a classical public key infrastructure
+
+Conclave uses Conclave Mail to connect clients to enclaves because of the following reasons:
+
+1. Compared to HTTPS, Conclave mail has a relatively small [Trusted Computing Base (TCB)](https://en.wikipedia.org/wiki/Trusted_computing_base).
```
Comment (shamsasari) [inline docs/docs/mail.md:146]: TLS, not HTTPS

File: docs/docs/mail.md [hunk 13]
```diff
@@ -1,345 +1,186 @@
...
-All this suggests the primary paradigm exposed to enclaves should be messages. The host will take care of
-delivering messages and providing atomic, transactional semantics over them so many kinds of enclaves won't need an
-encrypted database at all. The untrusted host can also take on the significant burden of moving data around,
-storing it to disk, sorting it, applying backpressure, exposing to the admin if there are backlogs etc.
+### Side-channel attacks
+
+A malicious host can infer crucial information if it knows the size or timing of a message.
+Conclave guards against these types of [side-channel attacks](security.md#side-channel-attacks) as follows.
+
+Conclave pads Mail items to a uniform size. This blocks any attempt to infer the contents of messages based on the
```
Comment (shamsasari) [inline docs/docs/mail.md:134]: Can we not use the phrase "item"? "Message" is the correct phrase.

File: docs/docs/mail.md [hunk 14]
```diff
@@ -1,345 +1,186 @@
...
+   send remote attestation to the client by returning it in another API, publishing it on a web server, and putting it
+   into a distributed hash table, network drive, or message queue.
+
+   Following are the benefits of the Noise protocol over an HTTPS/REST architecture:
+
+   * Eliminate complexity from enclave code.
+   * Avoid hacks like pseudo-certificates.
+   * Enable enclaves to talk to each other even if they can't be loaded simultaneously.
+   * Move session management *and expiry* out of the enclave.
+   * Get other benefits like reusing MQ brokers, integrating with Corda flows, storing messages to databases, and
```
Comment (shamsasari) [inline docs/docs/mail.md:178]: Delete this last bullet point.

File: docs/docs/mail.md (hunk 9 above)
Comment (JohnJoser3) [reply docs/docs/mail.md:64]: Corrected in [another PR](https://github.com/R3Conclave/conclave-core-sdk/pull/58).

File: docs/docs/mail.md (hunk 10 above)
Comment (JohnJoser3) [reply docs/docs/mail.md:127]: Corrected in [another PR](https://github.com/R3Conclave/conclave-core-sdk/pull/58).

File: docs/docs/mail.md (hunk 11 above)
Comment (JohnJoser3) [reply docs/docs/mail.md:7]: I have split this section into two to reduce the cognitive load on the user. The first section covers the components of a Mail item, and the second covers the internal working. 

As this doc is more of a reference material than a tutorial, I believe it's better to have these sections before describing the features of Conclave. The internal workings section adds more context to the features section and other sections below.

File: docs/docs/mail.md (hunk 14 above)
Comment (JohnJoser3) [reply docs/docs/mail.md:178]: Corrected in [another PR](https://github.com/R3Conclave/conclave-core-sdk/pull/58).

**Pull request merged**